src/java.base/windows/native/libjava/FileDescriptor_md.c

```diff
@@ -94,6 +97,54 @@ Java_java_io_FileCleanable_cleanupClose0(JNIEnv *env, jclass fdClass, jint unuse
 }
 
 JNIEXPORT jstring JNICALL
-Java_java_io_FileDescriptor_nativeDescription0(JNIEnv *env, jobject this) {
-    return (*env)->NewStringUTF(env, "(not implemented)");
-}
\ No newline at end of file
+Java_java_io_FileDescriptor_nativeDescription0(JNIEnv* env, jobject this) {
+    HANDLE handle = (HANDLE)(*env)->GetLongField(env, this, IO_handle_fdID);
+    #define BufferSize 1024
+    char lpszFilePath[BufferSize] = {'\0'};
+
+    HMODULE hKernelDll = LoadLibrary(TEXT("kernel32.dll"));
+    if (!hKernelDll) {
+        JNU_ThrowIOExceptionWithLastError(env, "LoadLibrary kernel32.dll failed");
+        return NULL;
+    }
+
+    HMODULE hNtdllDll = GetModuleHandle(TEXT("ntdll.dll"));
+    if (!hNtdllDll) {
+        JNU_ThrowIOExceptionWithLastError(env, "LoadLibrary ntdll.dll failed");
+        CloseHandle(hKernelDll);
+        return NULL;
+    }
+
+    typedef NTSTATUS(WINAPI* NtQueryObjectFunc)(HANDLE, OBJECT_INFORMATION_CLASS, PVOID, ULONG, PULONG);
```

Comment:
Rather than using a non-public function, would `GetFileInformationByHandle` work here? You can just test the file attributes matching https://learn.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants

Reply:
GetFileInformationByHandle doesn't work for sockets. The closest appropriate function would be GetFileType, however it operates with integer file type codes, and has too few codes to distiguidh types. Additionally, since I'd like to dump type name into a native description string, I have taken NtQueryObject.

Reply:
What do you mean by 'not working'? From what I see, you don't really use the function except for identifying that it is really a file/dir. Therefore if it returns an error for a socket, you can immediately fall back to the `Handle %p`. Or does the NtQueryObject actually provide any info about the socket that you read?

Reply:
I thought it'd be useful to have a name of a handle type while creating a native description string. That's why I used NtQueryObject. However it's just an info string, so I agree that using NtQueryObject is not necessary here.
I will replace it with GetFileType.

Reply:
I must have missed it, because I haven't seen the handle type being reported anywhere.